### PR TITLE
[lit-html] Adds element bindings and ref() directive

### LINKED ISSUE #923: Add syntax for spreading attributes, properties, and event listeners
It would be nice to be able to spread bindings in lit-html. One would be able to spread properties, attributes, and event listeners.

Syntax could look something like this:
```javascript
const spreadable = { '.property': 1, 'attribute': 'foo', '@click': handler }
html`<div ...=${spreadable}>How convenient</div>`
```

This feature request was discussed at length in this pull request: https://github.com/Polymer/lit-html/pull/213

We're investigating the tradeoff between convenience and bloat associated with this feature. This issue is for feature tracking.

### PR DESCRIPTION
**Pull request by justinfagnani**
This doesn't fully fix #923, but a spread directive can be built on top of this.

### CODE REVIEW

File: packages/lit-html/src/lib/directives/ref.ts
Comment (kevinpschaaf) [inline packages/lit-html/src/lib/directives/ref.ts:37]: Pattern feedback: It's odd to me to have a `render` method but never actually call it from `update`.

File: packages/lit-html/src/lib/lit-html.ts
Comment (kevinpschaaf) [inline packages/lit-html/src/lib/lit-html.ts:936]: Should maybe remove the previous value's bound attributes (if value was of type `AttributesTemplate`)? Should probably do that regardless, if the type of the previous value didn't match new value. This is the tricky part.

**Review by kevinpschaaf — COMMENTED** [review]
So, we need to decide what we're going to do overallwith SpreadPart/AttributeTemplate. To me, the only sane thing is to factor NodePart/Template(/Instance) so that spread can reuse the code related to parsing, creating, and committing AttributePart/EventPart/PropertyParts.

I don't think we want a dichotomy of syntax support or duplication of code here. Are there other approaches you're considering?

File: packages/lit-html/src/lib/lit-html.ts
Comment (sorvell) [inline packages/lit-html/src/lib/lit-html.ts:928]: It's unfortunate that each part type has to deal with directives. Can this be centralized?

File: packages/lit-html/src/lib/lit-html.ts
Comment (sorvell) [inline packages/lit-html/src/lib/lit-html.ts:966]: What about property and event bindings and values that are not strings? This feature doesn't seem worth it unless you can bind what you normally can in attribute position.

**Review by sorvell — CHANGES_REQUESTED** [review]

File: packages/lit-html/src/directives/ref.ts
Comment (arthurevans) [inline packages/lit-html/src/directives/ref.ts:51]:
```suggestion
 *     inputRef.value.focus();
```

File: packages/lit-html/src/directives/ref.ts
Comment (arthurevans) [inline packages/lit-html/src/directives/ref.ts:50]:
```suggestion
 *     render(html`<input ${ref(inputRef)}>`, container);
```

File: packages/lit-html/src/lib/directives/ref.ts
Comment (justinfagnani) [reply packages/lit-html/src/lib/directives/ref.ts]: Right now `render()` is needed to generate the directive function signature. We can try to update the types to use `update()` if `render()` isn't defined.

File: packages/lit-html/src/lib/lit-html.ts
Comment (justinfagnani) [reply packages/lit-html/src/lib/lit-html.ts]: N/A since removing attr templates

File: packages/lit-html/src/lib/lit-html.ts
Comment (justinfagnani) [reply packages/lit-html/src/lib/lit-html.ts]: N/A

File: packages/lit-html/src/lib/lit-html.ts
Comment (justinfagnani) [reply packages/lit-html/src/lib/lit-html.ts]: Kevin has part of this in #1426

File: packages/lit-html/src/lib/lit-html.ts
Comment (justinfagnani) [reply packages/lit-html/src/lib/lit-html.ts]: N/A

File: packages/lit-html/src/lib/directives/ref.ts
Comment (justinfagnani) [reply packages/lit-html/src/lib/directives/ref.ts]: I think that makes common usage a little more cumbersome. You'll always need a way to read a ref, so you need some container.

For a common case of linking two elements:

```ts
const fooRef = new Ref();
render(html`
  <x-foo ${ref(fooRef)}></x-foo>
  <x-bar .foo=${fooRef.value}></x-bar>
`, container);
```

you'd have to do:
```ts
const foo;
const fooRef (e) => fooRefValue = e;
render(html`
  <x-foo ${ref(fooRef)}></x-foo>
  <x-bar .foo=${foo}></x-bar>
`, container);
```

Not hugely different, but enough I think. I'm also looking forward towards possibly having a RefController that can trigger a re-render if the element changes. This would have the same `set()`/`value` API. React can take both Ref objects and callbacks, btw. We can discuss.

File: packages/lit-html/src/directives/ref.ts
Comment (justinfagnani) [reply packages/lit-html/src/directives/ref.ts:50]: done

File: packages/lit-html/src/directives/ref.ts
Comment (justinfagnani) [reply packages/lit-html/src/directives/ref.ts:49]: done

**Review by sorvell — CHANGES_REQUESTED** [review]

**Comment by sorvell** [comment]
Notes from offline chat:

- Right now the template cache is only keyed on the static strings in the template and this means the following type of usage is not supported:

  ```
  inputTemplate(type) => html`<input type=${unsafeStatic(type)}>` // always returns first call result
  ```
  We do want to support this use case. To do so, the static values must be included in the template key.

- To avoid confusion between `ref` and `Ref`, we'll change `new Ref()` to `createRef()`.
- To avoid leaking,`ref` should become a disconnectable directive and clear its value when disconnected. We *might* want to add an option to avoid disposing of the ref.
- We want to support easily binding a ref node into a property of another element in the template. This works today only if the accepting element is after the node where the ref is established in the template. We'd like to support this for elements before the ref in the template. We'll probably tackle this in a subsequent PR but it might be something like `forwardRef(fooRef)` and `forwardRef` can know when `fooRef` is set and either set its value or trigger an update on an enclosing LitElement (via controller integration).

File: packages/lit-html/src/test/directives/ref_test.ts
Comment (justinfagnani) [reply packages/lit-html/src/test/directives/ref_test.ts:32]: done

File: packages/lit-html/src/lit-html.ts
Comment (justinfagnani) [reply packages/lit-html/src/lit-html.ts:1341]: Possibly with the factoring out of resolveDirective, though this is more streamlined since we only accept directives here.

File: packages/lit-html/src/lit-html.ts
Comment (justinfagnani) [reply packages/lit-html/src/lit-html.ts:547]: (3) had a sub-case. I separated it out

File: packages/lit-html/src/lit-html.ts
Comment (justinfagnani) [reply packages/lit-html/src/lit-html.ts:550]: fixed

File: packages/lit-html/src/directives/ref.ts
Comment (justinfagnani) [reply packages/lit-html/src/directives/ref.ts:33]: it's not... this isn't observable yet, but when we add `forwardingRef`, we'll need something like this.

File: packages/lit-html/src/directives/ref.ts
Comment (justinfagnani) [reply packages/lit-html/src/directives/ref.ts:22]: When the set becomes observable with `forwardingRef` (or whatever it's called) I want the method to indicate that this can cause work.

File: packages/lit-html/src/test/directives/ref_test.ts
Comment (justinfagnani) [reply packages/lit-html/src/test/directives/ref_test.ts:40]: Added a conditional rendering test which uses a ref on more than one element, but only one at a time.

We need to add other tests when we get disconnectable directives

File: packages/lit-html/src/directives/ref.ts
Comment (sorvell) [inline packages/lit-html/src/directives/ref.ts:55]:
```suggestion
 * callback is a function that takes an element as its only argument.
```

File: packages/lit-html/src/directives/ref.ts
Comment (sorvell) [inline packages/lit-html/src/directives/ref.ts:43]: If you pass a different ref object, it'll always be filled, but if you pass a different ref callback it won't. While a corner case, that's unexpected. Seems like we should either always call the ref function or store and dirty check it.

File: packages/lit-html/src/lit-html.ts
Comment (sorvell) [reply packages/lit-html/src/lit-html.ts:1344]: ok

File: packages/lit-html/src/test/directives/ref_test.ts
Comment (sorvell) [inline packages/lit-html/src/test/directives/ref_test.ts:77]: Add tests for changing the ref object/callback passed to `ref`, or let's specifically decide not to support that.

File: packages/lit-html/src/directives/ref.ts
Comment (sorvell) [reply packages/lit-html/src/directives/ref.ts:43]: Suggest we decide whether or not to support changing what you pass to `ref` and handling the callback/object cases the same.

**Review by sorvell — CHANGES_REQUESTED** [review]

**Review by kevinpschaaf — COMMENTED** [review]
Code looks good with exception of handling Ref callback changing correctly as Steve noted.

Let's also either hold this PR or add issues for integrating with these pending changes:
* Use `resolveDirective` helper in `ElementPart._setValue` (once #1426 lands)
* Make `ref` a `DisconnectableDirective` to handle clearing the box (once #1432 lands)

File: packages/lit-html/src/directives/ref.ts
Comment (justinfagnani) [inline packages/lit-html/src/directives/ref.ts]: Updated to track the identity of the Ref or callback

File: packages/lit-html/src/directives/ref.ts
Comment (justinfagnani) [reply packages/lit-html/src/directives/ref.ts]: fixed

File: packages/lit-html/src/test/directives/ref_test.ts
Comment (justinfagnani) [reply packages/lit-html/src/test/directives/ref_test.ts:148]: done

**Comment by sorvell** [comment]
Do we need to specifically add a forwardRef or is it just adding tests that bind the ref down? And if it's the latter, let's just add those tests?

**Review by sorvell — CHANGES_REQUESTED** [review]
If forwarding refs just needs tests, let's add those?

Let's also merge the other directive stuff before merging this.

File: packages/lit-html/src/test/directives/ref_test.ts
Comment (sorvell) [inline packages/lit-html/src/test/directives/ref_test.ts:153]: Add a test to verify the refs are always set in tree order:

```
html`<div id="first" ${ref(myRef)}></div><div id="next" ${ref(myRef)}>${html`<span id="last" ${ref(myRef)}`}</div>`

File: packages/lit-html/src/directives/ref.ts
Comment (sorvell) [reply packages/lit-html/src/directives/ref.ts]: We elected not to add `fowardingRef` since this can be covered by using a callback.

File: packages/lit-html/src/directives/ref.ts
Comment (sorvell) [inline packages/lit-html/src/directives/ref.ts:35]: Not sure why we have this if we're not doing `fowardingRef`.

File: packages/lit-html/src/lit-html.ts
Comment (sorvell) [inline packages/lit-html/src/lit-html.ts:1475]: Can the actual code here be in the `Part` base class to be more DRY?

**Review by sorvell — CHANGES_REQUESTED** [review]

File: packages/lit-html/src/lit-html.ts
Comment (kevinpschaaf) [reply packages/lit-html/src/lit-html.ts:1475]: I have a stash where I tried making `ElementPart` a base class to `AttributePart` (since it's basically a superset of it), but it still needs a separate class, `_$setValue` implementation, and `AttributePart` then needs to super to it, so the checksize didn't change meaningfully (to the number of digits it shows).

Comment (sorvell) [reply packages/lit-html/src/lit-html.ts:1475]: ok

File: packages/lit-html/src/directives/ref.ts
Comment (justinfagnani) [reply packages/lit-html/src/directives/ref.ts:35]: My thought was we could eventually make Ref observable to trigger re-rendering when you forwarded it. We can do that later though, and with a setter.

File: packages/lit-html/src/test/directives/ref_test.ts
Comment (kevinpschaaf) [reply packages/lit-html/src/test/directives/ref_test.ts:153]: Good call. This is a weird usage of ref because it's last-in-wins (aka there's no point in 'first' and 'next' refs since they'll always be overwritten), but we weren't calling callbacks correctly in this setup (if `myRef` was a callback, it should call all 3 in order each render I think...?). Added tests for that.

File: packages/lit-html/src/directives/ref.ts
Comment (kevinpschaaf) [reply packages/lit-html/src/directives/ref.ts:35]: Removed for now and made `readonly`.

**Comment by kevinpschaaf** [comment]
@justinfagnani @sorvell Updated based on feedback, PTAL

**Review by sorvell — APPROVED** [review]
LGTM

File: packages/lit-html/src/directives/ref.ts
Comment (justinfagnani) [inline packages/lit-html/src/directives/ref.ts:71]: I'd leave a little comment why we do this. There are cases where we can't _not_ call with undefined when the ref changes and this is to normalize all the cases, yeah?

**Review by justinfagnani — COMMENTED** [review]
LGTM

File: packages/lit-html/src/directives/ref.ts
Comment (kevinpschaaf) [reply packages/lit-html/src/directives/ref.ts:71]: Exactly. I'll clarify.

**Pull request merged**
